modules/c++/sys/source/OSUnix.cpp

```diff
@@ -132,13 +133,68 @@ std::set<std::string> get_unique_thread_siblings()
     return unique_ts;
 }
 
-void get_cpu_affinity(cpu_set_t* mask)
+class ScopedCPUAffinity
 {
-    if (-1 == sched_getaffinity(0, sizeof(cpu_set_t), mask))
+public:
+    ScopedCPUAffinity()
     {
-        throw except::Exception(Ctxt("Failed to get CPU affinity"));
+        int numOnlineCPUs = sysconf(_SC_NPROCESSORS_ONLN);
+        if (numOnlineCPUs == -1)
+        {
+            throw except::Exception(Ctxt("Failed to get online CPU count"));
+        }
+
+        // The number of processors in the mask must be _at least_ the number of
+        // CPU's representable by the kernel's internal mask. This is given by
+        // the constant CPU_SETSIZE.
+        int maxCPUs = std::max(numOnlineCPUs, CPU_SETSIZE);
+        mSize = CPU_ALLOC_SIZE(maxCPUs);
+        mMask = CPU_ALLOC(maxCPUs);
```

Comment:
So this'll actually leak if the check below `sched_getaffinity()` throws because you'll never get into the destructor.  So you need an RAII class just dealing with this allocation/freeing.

Reply:
Worth it to have a whole new class, or just a

``` 
if (mMask != NULL)
{
    CPU_FREE(mMask);
}
```

inside the failure if-statement?

Reply:
Let's have a whole new class to let RAII auto-magically handle everything for us